Foundation/NSPathUtilities.swift

```diff
@@ -82,15 +82,15 @@ internal extension String {
     }
     
     internal func _stringByAppendingPathComponent(_ str: String, doneAppending : Bool = true) -> String {
-        if str.length == 0 {
-            return self
-        }
-        if self == "" {
-            return "/" + str
-        }
-        if self == "/" {
+        if hasSuffix("/") {
```

Comment:
I think the isEmpty check should become before the self check, as it was before. Is there a reason why they were reversed?

The change to use hasSuffix is a good one, though, and will work in the self == "/" case.

Reply:
> I think the isEmpty check should become before the self check, as it was before. Is there a reason why they were reversed?

I thought about it and I think appending a path component to an empty string is less common than appending to actual path, therefore, cases where `isEmpty` evaluates true should be fewer. Most cases will be handled in `hasSuffix` and last line of the function ( `self + "/" + str` )

What do you think about it?

Reply:
That may be the case, but we now have different short-circuit behaviour on both platforms, and it's really unrelated to this change (which is to fix the bug where appending something onto a string that ends with / results in two // I suspect). I would prefer it to keep it as is (and minimise diffs) in order to fix this bug, and then if it is decided that the re-ordering of the test cases makes sense then it can be done. Several bugs have made it through to production where code was rewritten as:

```
a->something();
if (a==NULL) { abort(); }
```

because a compiler could 'prove' that dereferencing something implied that the variable could never be null, since otherwise would be undefined behaviour and thus the if check could be compiled out completely from the code.

While Swift doesn't suffer from this, the `isEmtpy` check could easily be handling such special-cases (now or in the future) and the re-ordering based on 'a thought' probably isn't the right way to tackle this issue.

Can you commit without the re-ordering on the checks so that you're just fixing the bug, and then squash all the commits so that we don't have interim mess getting in the way? We can then propose via separate PR/bug report that reordering the cases is desirable, with data to back up that particular part of the claim.

Reply:
@alblue  ok, what about calling Swift.String's internal function from NSString.append... ?

Reply:
I think it makes sense for the calls to be de-duplicated as you have done now; but @parkera or @phausler are the ones to definitively make that call. For me, the less duplicated code we have, the better :)